### ISSUE TITLE
Value with select for source input is not valid 

### ISSUE THREAD

**Issue by rldhont** [3liz/py-qgis-wps#10]
To run an algorithm on a part of a layer, it is possible to add a `select` param to the layer name:
* [layersio.py#L244](https://github.com/3liz/py-qgis-wps/blob/master/pyqgiswps/executors/io/layersio.py#L244)
* [test_algorithms.py#L303](https://github.com/3liz/py-qgis-wps/blob/master/tests/unittests/processing/test_algorithms.py#L303)

But if you do so, py-qgis-wps returns `Exception: code: 400, description: Input data not valid using mode 1`
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!-- PyQgisWPS 1.4.3 -->
<ows:ExceptionReport xmlns:ows="http://www.opengis.net/ows/1.1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.opengis.net/ows/1.1 http://schemas.opengis.net/ows/1.1.0/owsExceptionReport.xsd" version="1.0.0">
  <ows:Exception exceptionCode="InvalidParameterValue" locator="" >
      <ows:ExceptionText>Input data not valid using mode 1</ows:ExceptionText>
  </ows:Exception>
</ows:ExceptionReport>
```

**Comment by dmarteau** [comment]
Can you give the server logs  ?

**Comment by dmarteau** [comment]
The part of the wps server logs that contains the request and the backtrace of the error.

**Comment by rldhont** [comment]
```
wps_1     | 2020-10-09 13:57:32,952	[1]	DEBUG	Request: execute
wps_1     | 2020-10-09 13:57:32,954	[1]	DEBUG	Checking if datainputs is required and has been passed
wps_1     | 2020-10-09 13:57:32,954	[1]	DEBUG	Checking if all mandatory inputs have been passed
wps_1     | 2020-10-09 13:57:32,954	[1]	DEBUG	Checking input: INPUT
wps_1     | 2020-10-09 13:57:32,954	[1]	DEBUG	validating allowed values: layer:france_parts?select=%24id%20IN%20(%201%2C2%20) in [AllowedValue(minval=None, maxval=None, range_closure=closed), AllowedValue(minval=None, maxval=None, range_closure=closed), AllowedValue(minval=None, maxval=None, range_closure=closed)]
wps_1     | 2020-10-09 13:57:32,954	[1]	DEBUG	validation result: False
wps_1     | 2020-10-09 13:57:32,954	[1]	ERROR	Exception: code: 400, description: Input data not valid using mode 1, locator: 
wps_1     | 2020-10-09 13:57:32,954	[1]	DEBUG	Request failed wint message: Bad Request HTTP 400: Bad Request
wps_1     | 2020-10-09 13:57:32,955	[1]	DEBUG	Serializing XML response
wps_1     | 2020-10-09 13:57:32,955	[1]	REQ	172.21.0.5	400	POST	/ows/?request=Execute&service=WPS&version=1%2E0%2E0&map=test%5Ffrance%5Fparts%2Ffrance%5Fparts%2Eqgs	5	499	
```

**Comment by rldhont** [comment]
The Request
```xml
<?xml version="1.0" encoding="UTF-8"?>
<wps:Execute xmlns:wps="http://www.opengis.net/wps/1.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="1.0.0" service="WPS" xsi:schemaLocation="http://www.opengis.net/wps/1.0.0 http://schemas.opengis.net/wps/1.0.0/wpsAll.xsd">
   <ows:Identifier xmlns:ows="http://www.opengis.net/ows/1.1">pyqgiswps_test:qgisbuffer</ows:Identifier>
   <wps:DataInputs>
      <wps:Input>
         <ows:Identifier xmlns:ows="http://www.opengis.net/ows/1.1">INPUT</ows:Identifier>
         <ows:Title xmlns:ows="http://www.opengis.net/ows/1.1">INPUT</ows:Title>
         <wps:Data>
            <wps:LiteralData>layer:france_parts?select=%24id%20IN%20(%201%2C2%20)</wps:LiteralData>
         </wps:Data>
      </wps:Input>
      <wps:Input>
         <ows:Identifier xmlns:ows="http://www.opengis.net/ows/1.1">DISTANCE</ows:Identifier>
         <ows:Title xmlns:ows="http://www.opengis.net/ows/1.1">DISTANCE</ows:Title>
         <wps:Data>
            <wps:LiteralData>0</wps:LiteralData>
         </wps:Data>
      </wps:Input>
      <wps:Input>
         <ows:Identifier xmlns:ows="http://www.opengis.net/ows/1.1">SEGMENTS</ows:Identifier>
         <ows:Title xmlns:ows="http://www.opengis.net/ows/1.1">SEGMENTS</ows:Title>
         <wps:Data>
            <wps:LiteralData>5</wps:LiteralData>
         </wps:Data>
      </wps:Input>
      <wps:Input>
         <ows:Identifier xmlns:ows="http://www.opengis.net/ows/1.1">END_CAP_STYLE</ows:Identifier>
         <ows:Title xmlns:ows="http://www.opengis.net/ows/1.1">END CAP STYLE</ows:Title>
         <wps:Data>
            <wps:LiteralData>Round</wps:LiteralData>
         </wps:Data>
      </wps:Input>
      <wps:Input>
         <ows:Identifier xmlns:ows="http://www.opengis.net/ows/1.1">JOIN_STYLE</ows:Identifier>
         <ows:Title xmlns:ows="http://www.opengis.net/ows/1.1">JOIN STYLE</ows:Title>
         <wps:Data>
            <wps:LiteralData>Round</wps:LiteralData>
         </wps:Data>
      </wps:Input>
      <wps:Input>
         <ows:Identifier xmlns:ows="http://www.opengis.net/ows/1.1">MITER_LIMIT</ows:Identifier>
         <ows:Title xmlns:ows="http://www.opengis.net/ows/1.1">MITER LIMIT</ows:Title>
         <wps:Data>
            <wps:LiteralData>2</wps:LiteralData>
         </wps:Data>
      </wps:Input>
      <wps:Input>
         <ows:Identifier xmlns:ows="http://www.opengis.net/ows/1.1">DISSOLVE</ows:Identifier>
         <ows:Title xmlns:ows="http://www.opengis.net/ows/1.1">DISSOLVE</ows:Title>
         <wps:Data>
            <wps:LiteralData>True</wps:LiteralData>
         </wps:Data>
      </wps:Input>
      <wps:Input>
         <ows:Identifier xmlns:ows="http://www.opengis.net/ows/1.1">OUTPUT</ows:Identifier>


         <ows:Title xmlns:ows="http://www.opengis.net/ows/1.1">OUTPUT</ows:Title>
         <wps:Data>
            <wps:LiteralData>buff</wps:LiteralData>
         </wps:Data>
      </wps:Input>
   </wps:DataInputs>
   <wps:ResponseForm>
      <wps:ResponseDocument storeExecuteResponse="true" status="true">
         <wps:Output asReference="true">
            <ows:Identifier xmlns:ows="http://www.opengis.net/ows/1.1">OUTPUT</ows:Identifier>
            <ows:Title xmlns:ows="http://www.opengis.net/ows/1.1" />
            <ows:Abstract xmlns:ows="http://www.opengis.net/ows/1.1" />
         </wps:Output>
      </wps:ResponseDocument>
   </wps:ResponseForm>
</wps:Execute>
```

**Comment by dmarteau** [comment]
The error does not come from the processing handling of the parameters buf from the WPS validation itself.  Can you give the DescribeProcess response ? It may gives some hint abount allowed values for the layer spec.

**Comment by rldhont** [comment]
```xml
      <Input minOccurs="1" maxOccurs="1">
        <ows:Identifier>INPUT</ows:Identifier>
        <ows:Title>INPUT</ows:Title>
        <ows:Abstract>Input layer</ows:Abstract>
        <ows:Metadata xlink:title="processing:type" xlink:href="source" xlink:type="simple"/>
        <ows:Metadata xlink:title="processing:dataTypes" xlink:href="TypeVector" xlink:type="simple"/>
        <LiteralData>
          <ows:DataType ows:reference="urn:ogc:def:dataType:OGC:1.1:string">string</ows:DataType>
          <ows:AllowedValues>
            <ows:Value>france_parts</ows:Value>
            <ows:Value>france_parts bordure</ows:Value>
            <ows:Value>france_parts tuil&#233; en cache</ows:Value>
          </ows:AllowedValues>
        </LiteralData>
      </Input>
```
If the WPS checks the AllowedValues, I can't use selection

**Comment by dmarteau** [comment]
The allowed values are a feature added for returning the list of available layers instead letting the client guessing the input layer names. This is clearly incompatible with passing url-like passing extra parameter along with the name.

**Comment by dmarteau** [comment]
May be we could change that and set the selection as  metadata of the input parameter ?

**Comment by dmarteau** [comment]
It is clear that we cannot  use `select` expression as query-like parameters attached to the layer name since it conflicts with the validation for layers input.
@rldhont: this means that the lizmap wps module must be changed to supports an alternative way of passing 'expression' or other metadata with the parameters.

**Comment by dmarteau** [comment]
The preferred solution is to validate allowed values as `layer` value type.

Note that the `layer:` scheme prefix is mandatory for `select` expressions beeing considered.

**Comment by dmarteau** [comment]
@rldhont that's mean there is nothing to do from Lizmap point of view.

**Comment by dmarteau** [comment]
Fixed in 1.4.4 - commit 5f17929 -

**Comment by rldhont** [comment]
Thanks @dmarteau
